plugins/modem-manager/fu-firehose-updater.c

```diff
@@ -0,0 +1,748 @@
+		GByteArray *cmd_bytearray = NULL;
+		g_autoptr(XbSilo) rsp_silo = NULL;
+		g_autoptr(XbNode) rsp_node = NULL;
+		GString *cmd_str = g_string_new (NULL);
+
+		g_string_append_printf (cmd_str, "<configure");
+		g_string_append_printf (cmd_str, " MemoryName=\"%s\"",
+					CONFIGURE_MEMORY_NAME);
+		g_string_append_printf (cmd_str, " Verbose=\"%d\"",
+					CONFIGURE_VERBOSE);
+		g_string_append_printf (cmd_str, " AlwaysValidate=\"%d\"",
+					CONFIGURE_ALWAYS_VALIDATE);
+		g_string_append_printf (cmd_str, " MaxDigestTableSizeInBytes=\"%d\"",
+					CONFIGURE_MAX_DIGEST_TABLE_SIZE_IN_BYTES);
+		g_string_append_printf (cmd_str, " MaxPayloadSizeToTargetInBytes=\"%d\"",
+					max_payload_size);
+		g_string_append_printf (cmd_str, " ZlpAwareHost=\"%d\"",
+					CONFIGURE_ZLP_AWARE_HOST);
+		g_string_append_printf (cmd_str, " SkipStorageInit=\"%d\"",
+					CONFIGURE_SKIP_STORAGE_INIT);
+		g_string_append_printf (cmd_str, "/>");
+
+		cmd_bytearray = g_bytes_unref_to_array (g_string_free_to_bytes (cmd_str));
+
+		if (!fu_firehose_updater_send_and_receive (self, cmd_bytearray, &rsp_silo, &rsp_node, error)) {
+			g_prefix_error (error, "Failed to run configure command: ");
+			return 0;
+		}
+
+		/* retry if we're told to use a different max payload size */
+		if (!fu_firehose_updater_check_operation_result (rsp_node, NULL)) {
+			guint64 suggested_max_payload_size;
+			g_autoptr(XbNode) root = NULL;
+
+			root = xb_silo_get_root (rsp_silo);
+			suggested_max_payload_size = xb_node_get_attr_as_uint (root, "MaxPayloadSizeToTargetInBytes");
+			if ((suggested_max_payload_size > G_MAXINT) ||
+			    ((gint)suggested_max_payload_size == max_payload_size)) {
+				break;
+			}
+
+			suggested_max_payload_size = max_payload_size;
+			continue;
+		}
+
+		/* if operation is successful, return the max payload size we requested */
+		return max_payload_size;
+	}
+
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "Configure operation failed");
+	return 0;
+}
+
+static gboolean
+fu_firehose_updater_reset (FuFirehoseUpdater *self, GError **error)
+{
+	guint recv_cnt = 20;
+	const gchar *cmd_str = "<power value=\"reset\" />";
+	GByteArray *cmd_bytearray = NULL;
+	g_autoptr(XbSilo) rsp_silo = NULL;
+	g_autoptr(XbNode) rsp_node = NULL;
+
+	cmd_bytearray = g_byte_array_append (g_byte_array_new (), (const guint8 *)cmd_str, strlen (cmd_str));
+	if (!fu_firehose_updater_send_and_receive (self, cmd_bytearray, &rsp_silo, &rsp_node, error)) {
+		g_prefix_error (error, "Failed to run reset command: ");
+		return FALSE;
+	}
+
+	if (!fu_firehose_updater_check_operation_result (rsp_node, NULL)) {
+		g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+				     "Reset operation failed");
+		return FALSE;
+	}
+
+	/* read out all of the remaining messages. otherwise modem won't go into reset */
+	while (--recv_cnt && fu_firehose_updater_send_and_receive (self, NULL, &rsp_silo, &rsp_node, error));
```

Comment:
if one of these send/receives fails you need to return false.

Reply:
it should fail actually. we are trying to read out all of the remaining messages in the modem's buffer, so we are reading until the send/receive function returns `FALSE`. that means modem is now ready to go into reset and function has finished successfully.
recv_cnt there is only to protect against infinite loop. that can only happen if something goes wrong with the modem. usually there are only 2 or 3 messages left.

i will add g_warn_if_fail (recv_cnt > 0) though

Reply:
If you're waiting and expecting until it returns false, you need a local error variable. Not the one passed.

Reply:
makes sense, will fix

Reply:
replaced with NULL, since it does not matter if that function returns error or not